app/client/components/GridView.js

```diff
@@ -403,6 +417,181 @@ GridView.prototype._shiftSelect = function(step, selectObs, exemptType, maxVal)
+  }
+
+  const colValuesByIndex = {};
+  for (const field of fields) {
+    const displayColId = field.displayColModel.peek().colId.peek();
+    colValuesByIndex[field._index()] = this.tableModel.tableData.getColValues(displayColId);
+  }
+
+  let steps = 0;
+
+  if (type === selector.ROW) {
+    const isLastColEmpty = this._areValuesInColEmpty(colEnd, rowIndices, colValuesByIndex);
+    const isNextColEmpty = this._areValuesInColEmpty(colEnd + direction, rowIndices, colValuesByIndex);
+    const shouldStopOnEmptyValue = !isLastColEmpty && !isNextColEmpty;
+    for (let i = 1; i < fields.length; i++) {
+      const hasEmptyValues = this._areValuesInColEmpty(fields[i]._index(), rowIndices, colValuesByIndex);
+      if (hasEmptyValues && shouldStopOnEmptyValue) {
+        return steps;
+      } else if (!hasEmptyValues && !shouldStopOnEmptyValue) {
+        return steps + 1;
+      }
+
+      steps += 1;
+    }
+  } else {
+    const isLastRowEmpty = this._areValuesInRowEmpty(rowIndices[0], colValuesByIndex);
+    const isNextRowEmpty = this._areValuesInRowEmpty(rowIndices[1], colValuesByIndex);
+    const shouldStopOnEmptyValue = !isLastRowEmpty && !isNextRowEmpty;
+    for (let i = 1; i < rowIndices.length; i++) {
+      const hasEmptyValues = this._areValuesInRowEmpty(rowIndices[i], colValuesByIndex);
+      if (hasEmptyValues && shouldStopOnEmptyValue) {
+        return steps;
+      } else if (!hasEmptyValues && !shouldStopOnEmptyValue) {
+        return steps + 1;
+      }
+
+      steps += 1;
+    }
+  }
+
+  return steps;
+}
+
+GridView.prototype._selectionData = function({colStart, colEnd, rowStart, rowEnd}) {
+  const fields = [];
+  for (let i = colStart; i <= colEnd; i++) {
+    const field = this.viewSection.viewFields().at(i);
+    if (!field) { continue; }
+
+    fields.push(field);
+  }
+
+  const rowIndices = [];
+  for (let i = rowStart; i <= rowEnd; i++) {
+    const rowId = this.viewData.getRowId(i);
+    if (!rowId) { continue; }
+
+    rowIndices.push(this.tableModel.tableData.getRowIdIndex(rowId));
+  }
+
+  return {fields, rowIndices};
+}
+
+GridView.prototype._areValuesInColEmpty = function(colIndex, rowIndices, colValuesByIndex) {
+  return rowIndices.every(rowIndex => {
+    const value = colValuesByIndex[colIndex][rowIndex];
+    return value === null || value === undefined || value === '' || value === 'false';
+  });
+}
+
+GridView.prototype._areValuesInRowEmpty = function(rowIndex, colValuesByIndex) {
+  return Object.values(colValuesByIndex).every((colValues) => {
+    const value = colValues[rowIndex];
+    return value === null || value === undefined || value === '' || value === 'false';
+  });
+}
+
+/**
+ * Returns a GridSelection of the given cell selection
+ *
+ * @param {integer} colStart
+ * @param {integer} colEnd
+ * @param {integer} rowStart
+ * @param {integer} rowEnd
+ * @returns {Object} CopySelection
+ */
+GridView.prototype.getCells = function(colStart, colEnd, rowStart, rowEnd) {
```

Comment:
`_getCopySelection` would be a slightly better name for this, I think.

Since all of the parameters are related, I'd suggest grouping them into a single object. We can then destructure it like so:

```javascript
GridView.prototype._getCopySelection = function({colStart, colEnd, rowStart, rowEnd}) {
```